File: code/0836/836合并集合.py
def find(x):
    if p[x] != x:
        p[x] = find(p[x])

    return p[x]


if __name__ == "__main__":
    n, m = map(int, input().split())
    p = [i for i in range(n + 1)]

    for _ in range(m):
        op = input().split()
        a, b = find(int(op[1])), find(int(op[2]))
        if op[0] == 'M':
            p[b] = a
        else:
            print("Yes") if a == b else print("No")

File: code/0836/836合并集合.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
using namespace std;

const int N = 1e5 + 10;
int p[N];
int n, m;

int find(int x) {
	if (p[x] != x)
		p[x] = find(p[x]);
	
	return p[x];
}

int main() {
	cin >> n >> m;
	for (int i = 1; i <= n; i++)
		p[i] = i;

	while (m--) {
		char op;
		int a, b;
		cin >> op >> a >> b;
		a = find(a), b = find(b);
		if (op == 'M') {
			if (a != b)
				p[a] = b;
		}
		else {
			if (a == b) puts("Yes");
			else puts("No");
		}
	}

	return 0;
}
